### ISSUE TITLE
Update fails

### ISSUE THREAD

**Issue by MidnightBolt** [Bukimedia/PrestaSharp#126]
Hello,

I'm trying to update a product but I get a permission (403) error. I have checked that the webservice key has permissions but still it fails. This is my code and below that the response. I don't know where to start investigating. It must be a setting or something on my prestashop server side. But what... Do you have any ideas? Or maybe a good place to ask this?

```c#

private void dgvProducts_CellValueChanged(object sender, DataGridViewCellEventArgs e)
{
    string BaseUrl = "https://www.______.com/api";
    string Account = "xxxxxxxxx";
    string Password = "";
    productFactory = new ProductFactory(BaseUrl, Account, Password);
    Bukimedia.PrestaSharp.Entities.product product = productFactory.Get(Convert.ToInt64( d.Rows[e.RowIndex][0]));
    product.name[0].Value = d.Rows[e.RowIndex][1].ToString();
    productFactory.Update(product);
}
```

```xml
Bukimedia.PrestaSharp.PrestaSharpException was unhandled
  HResult=-2146232832
  Message=
15

<?xml version="1.0" encoding="utf-8"?>
<prestashop>
<product>
  <id xmlns="">15</id>
  <id_supplier xmlns="">0</id_supplier>
  <id_manufacturer xmlns="">0</id_manufacturer>
  <id_category_default xmlns="">2</id_category_default>
  <id_default_image d2p1:nil="true" xmlns:d2p1="http://www.w3.org/2001/XMLSchema-instance" xmlns="" />
  <id_default_combination d2p1:nil="true" xmlns:d2p1="http://www.w3.org/2001/XMLSchema-instance" xmlns="" />
  <id_tax_rules_group xmlns="">58</id_tax_rules_group>
  <id_shop_default xmlns="">1</id_shop_default>
  <on_sale xmlns="">0</on_sale>
  <online_only xmlns="">0</online_only>
  <ean13 xmlns="" />
  <upc xmlns="" />
  <ecotax xmlns="">0.000000</ecotax>
  <minimal_quantity xmlns="">1</minimal_quantity>
  <price xmlns="">0.000000</price>
  <wholesale_price xmlns="">3.520000</wholesale_price>
  <unity xmlns="" />
  <unit_price_ratio xmlns="">0.000000</unit_price_ratio>
  <additional_shipping_cost xmlns="">0.00</additional_shipping_cost>
  <reference xmlns="" />
  <supplier_reference xmlns="" />
  <location xmlns="" />
  <width xmlns="">0.000000</width>
  <height xmlns="">0.000000</height>
  <depth xmlns="">0.000000</depth>
  <weight xmlns="">0.012000</weight>
  <quantity_discount xmlns="">0</quantity_discount>
  <customizable xmlns="">0</customizable>
  <uploadable_files xmlns="">0</uploadable_files>
  <text_fields xmlns="">0</text_fields>
  <active xmlns="">1</active>
  <redirect_type xmlns="" />
  <id_product_redirected xmlns="">0</id_product_redirected>
  <available_for_order xmlns="">1</available_for_order>
  <available_date xmlns="">0000-00-00</available_date>
  <condition xmlns="">new</condition>
  <show_price xmlns="">1</show_price>
  <indexed xmlns="">0</indexed>
  <visibility xmlns="">both</visibility>
  <cache_is_pack xmlns="">0</cache_is_pack>
  <cache_has_attachments xmlns="">0</cache_has_attachments>
  <is_virtual xmlns="">0</is_virtual>
  <cache_default_attribute xmlns="">0</cache_default_attribute>
  <date_add xmlns="">2015-02-08 12:03:35</date_add>
  <date_upd xmlns="">2015-02-08 12:03:41</date_upd>
  <advanced_stock_management xmlns="">0</advanced_stock_management>
  <meta_description xmlns="">
    <language id="2" />
    <language id="3" />
  </meta_description>
  <meta_keywords xmlns="">
    <language id="2" />
    <language id="3" />
  </meta_keywords>
  <meta_title xmlns="">
    <language id="2" />
    <language id="3" />
  </meta_title>
  <link_rewrite xmlns="">
    <language id="2">925-silver-plated-earrings-double-banana-shape-ear-drop-jewelry</language>
    <language id="3">925-silver-plated-earrings-double-banana-shape-ear-drop-jewelry</language>
  </link_rewrite>
  <name xmlns="">
    <language id="2">Silver Plated Earrings Double Banana Shape Ear Drop Jewelry</language>
    <language id="3">925 Silver Plated Earrings Double Banana Shape Ear Drop Jewelry</language>
  </name>
  <description xmlns="">
    <language id="2" />
    <language id="3" />
  </description>
  <description_short xmlns="">
    <language id="2" />
    <language id="3" />
  </description_short>
  <available_now xmlns="">
    <language id="2" />
    <language id="3" />
  </available_now>
  <available_later xmlns="">
    <language id="2" />
    <language id="3" />
  </available_later>
  <associations xmlns="">
    <categories xmlns="">
      <category>
        <id>12</id>
      </category>
      <category>
        <id>17</id>
      </category>
      <category>
        <id>19</id>
      </category>
    </categories>
    <images xmlns="" />
    <combinations xmlns="" />
    <product_option_values xmlns="" />
    <product_features xmlns="" />
    <tags xmlns="" />
    <stock_availables xmlns="">
      <stock_available>
        <id>15</id>
        <id_product_attribute>0</id_product_attribute>
      </stock_available>
    </stock_availables>
    <product_bundle xmlns="" />
    <accessories xmlns="" />
  </associations>
</product>
</prestashop>

IYD8BL9IWJVM2LIW7xxxxxxxxxxxxxxx

Basic SVlEOEJMOUlXSlZNMkxJVzdHRzc0MzZDxxxxxxxxxxxx

application/json, application/xml, text/json, text/x-json, text/javascript, text/xml

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>403 Forbidden</title>
</head><body>
<h1>Forbidden</h1>
<p>You don't have permission to access /api/products
on this server.</p>
<p>Additionally, a 404 Not Found
error was encountered while trying to use an ErrorDocument to handle the request.</p>
<hr>
<address>Apache/2 Server at www.ilsjuwelenshop.com Port 443</address>
</body></html>
 '>' is an unexpected token. The expected token is '"' or '''. Line 1, position 50. HttpStatusCode: Forbidden
  Source=Bukimedia.PrestaSharp
  ResponseContent=
15

<?xml version="1.0" encoding="utf-8"?>
<prestashop>
<product>
  <id xmlns="">15</id>
  <id_supplier xmlns="">0</id_supplier>
  <id_manufacturer xmlns="">0</id_manufacturer>
  <id_category_default xmlns="">2</id_category_default>
  <id_default_image d2p1:nil="true" xmlns:d2p1="http://www.w3.org/2001/XMLSchema-instance" xmlns="" />
  <id_default_combination d2p1:nil="true" xmlns:d2p1="http://www.w3.org/2001/XMLSchema-instance" xmlns="" />
  <id_tax_rules_group xmlns="">58</id_tax_rules_group>
  <id_shop_default xmlns="">1</id_shop_default>
  <on_sale xmlns="">0</on_sale>
  <online_only xmlns="">0</online_only>
  <ean13 xmlns="" />
  <upc xmlns="" />
  <ecotax xmlns="">0.000000</ecotax>
  <minimal_quantity xmlns="">1</minimal_quantity>
  <price xmlns="">0.000000</price>
  <wholesale_price xmlns="">3.520000</wholesale_price>
  <unity xmlns="" />
  <unit_price_ratio xmlns="">0.000000</unit_price_ratio>
  <additional_shipping_cost xmlns="">0.00</additional_shipping_cost>
  <reference xmlns="" />
  <supplier_reference xmlns="" />
  <location xmlns="" />
  <width xmlns="">0.000000</width>
  <height xmlns="">0.000000</height>
  <depth xmlns="">0.000000</depth>
  <weight xmlns="">0.012000</weight>
  <quantity_discount xmlns="">0</quantity_discount>
  <customizable xmlns="">0</customizable>
  <uploadable_files xmlns="">0</uploadable_files>
  <text_fields xmlns="">0</text_fields>
  <active xmlns="">1</active>
  <redirect_type xmlns="" />
  <id_product_redirected xmlns="">0</id_product_redirected>
  <available_for_order xmlns="">1</available_for_order>
  <available_date xmlns="">0000-00-00</available_date>
  <condition xmlns="">new</condition>
  <show_price xmlns="">1</show_price>
  <indexed xmlns="">0</indexed>
  <visibility xmlns="">both</visibility>
  <cache_is_pack xmlns="">0</cache_is_pack>
  <cache_has_attachments xmlns="">0</cache_has_attachments>
  <is_virtual xmlns="">0</is_virtual>
  <cache_default_attribute xmlns="">0</cache_default_attribute>
  <date_add xmlns="">2015-02-08 12:03:35</date_add>
  <date_upd xmlns="">2015-02-08 12:03:41</date_upd>
  <advanced_stock_management xmlns="">0</advanced_stock_management>
  <meta_description xmlns="">
    <language id="2" />
    <language id="3" />
  </meta_description>
  <meta_keywords xmlns="">
    <language id="2" />
    <language id="3" />
  </meta_keywords>
  <meta_title xmlns="">
    <language id="2" />
    <language id="3" />
  </meta_title>
  <link_rewrite xmlns="">
    <language id="2">925-silver-plated-earrings-double-banana-shape-ear-drop-jewelry</language>
    <language id="3">925-silver-plated-earrings-double-banana-shape-ear-drop-jewelry</language>
  </link_rewrite>
  <name xmlns="">
    <language id="2">Silver Plated Earrings Double Banana Shape Ear Drop Jewelry</language>
    <language id="3">925 Silver Plated Earrings Double Banana Shape Ear Drop Jewelry</language>
  </name>
  <description xmlns="">
    <language id="2" />
    <language id="3" />
  </description>
  <description_short xmlns="">
    <language id="2" />
    <language id="3" />
  </description_short>
  <available_now xmlns="">
    <language id="2" />
    <language id="3" />
  </available_now>
  <available_later xmlns="">
    <language id="2" />
    <language id="3" />
  </available_later>
  <associations xmlns="">
    <categories xmlns="">
      <category>
        <id>12</id>
      </category>
      <category>
        <id>17</id>
      </category>
      <category>
        <id>19</id>
      </category>
    </categories>
    <images xmlns="" />
    <combinations xmlns="" />
    <product_option_values xmlns="" />
    <product_features xmlns="" />
    <tags xmlns="" />
    <stock_availables xmlns="">
      <stock_available>
        <id>15</id>
        <id_product_attribute>0</id_product_attribute>
      </stock_available>
    </stock_availables>
    <product_bundle xmlns="" />
    <accessories xmlns="" />
  </associations>
</product>
</prestashop>

IYD8BxxxxxxxxxxxxG743xxxxxxxxxxxx

Basic SVlEOExxxxzdHRzc0MzZxxxxxQTI6

application/json, application/xml, text/json, text/x-json, text/javascript, text/xml

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>403 Forbidden</title>
</head><body>
<h1>Forbidden</h1>
<p>You don't have permission to access /api/products
on this server.</p>
<p>Additionally, a 404 Not Found
error was encountered while trying to use an ErrorDocument to handle the request.</p>
<hr>
<address>Apache/2 Server at www.ilsjuwelenshop.com Port 443</address>
</body></html>

  ResponseErrorMessage='>' is an unexpected token. The expected token is '"' or '''. Line 1, position 50.
  StackTrace:
       at Bukimedia.PrestaSharp.Factories.RestSharpFactory.Execute[T](RestRequest Request) in c:\Users\dennis\Documents\Visual Studio 2013\Projects\PrestaManager\PrestaSharp\Factories\RestSharpFactory.cs:line 54
       at Bukimedia.PrestaSharp.Factories.ProductFactory.Update(product Product) in c:\Users\dennis\Documents\Visual Studio 2013\Projects\PrestaManager\PrestaSharp\Factories\ProductFactory.cs:line 39
       at PrestaManager.frmMain.dgvProducts_CellValueChanged(Object sender, DataGridViewCellEventArgs e) in c:\Users\dennis\Documents\Visual Studio 2013\Projects\PrestaManager\PrestaManager\Form1.cs:line 68
       at System.Windows.Forms.DataGridView.OnCellValueChanged(DataGridViewCellEventArgs e)
       at System.Windows.Forms.DataGridView.OnCellValueChangedInternal(DataGridViewCellEventArgs e)
       at System.Windows.Forms.DataGridViewCell.SetValue(Int32 rowIndex, Object value)
       at System.Windows.Forms.DataGridView.PushFormattedValue(DataGridViewCell& dataGridViewCurrentCell, Object formattedValue, Exception& exception)
       at System.Windows.Forms.DataGridView.CommitEdit(DataGridViewCell& dataGridViewCurrentCell, DataGridViewDataErrorContexts context, DataGridViewValidateCellInternal validateCell, Boolean fireCellLeave, Boolean fireCellEnter, Boolean fireRowLeave, Boolean fireRowEnter, Boolean fireLeave)
       at System.Windows.Forms.DataGridView.EndEdit(DataGridViewDataErrorContexts context, DataGridViewValidateCellInternal validateCell, Boolean fireCellLeave, Boolean fireCellEnter, Boolean fireRowLeave, Boolean fireRowEnter, Boolean fireLeave, Boolean keepFocus, Boolean resetCurrentCell, Boolean resetAnchorCell)
       at System.Windows.Forms.DataGridView.CommitEditForOperation(Int32 columnIndex, Int32 rowIndex, Boolean forCurrentCellChange)
       at System.Windows.Forms.DataGridView.ScrollIntoView(Int32 columnIndex, Int32 rowIndex, Boolean forCurrentCellChange)
       at System.Windows.Forms.DataGridView.ProcessDownKeyInternal(Keys keyData, Boolean& moved)
       at System.Windows.Forms.DataGridView.ProcessEnterKey(Keys keyData)
       at System.Windows.Forms.DataGridView.ProcessDialogKey(Keys keyData)
       at System.Windows.Forms.Control.ProcessDialogKey(Keys keyData)
       at System.Windows.Forms.TextBoxBase.ProcessDialogKey(Keys keyData)
       at System.Windows.Forms.Control.PreProcessMessage(Message& msg)
       at System.Windows.Forms.Control.PreProcessControlMessageInternal(Control target, Message& msg)
       at System.Windows.Forms.Application.ThreadContext.PreTranslateMessage(MSG& msg)
       at System.Windows.Forms.Application.ThreadContext.System.Windows.Forms.UnsafeNativeMethods.IMsoComponent.FPreTranslateMessage(MSG& msg)
       at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
       at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
       at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
       at System.Windows.Forms.Application.Run(Form mainForm)
       at PrestaManager.Program.Main() in c:\Users\dennis\Documents\Visual Studio 2013\Projects\PrestaManager\PrestaManager\Program.cs:line 19
       at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
       at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
       at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
       at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
       at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
       at System.Threading.ThreadHelper.ThreadStart()
  InnerException: System.Xml.XmlException
       HResult=-2146232000
       Message='>' is an unexpected token. The expected token is '"' or '''. Line 1, position 50.
       Source=System.Xml
       LineNumber=1
       LinePosition=50
       StackTrace:
            at System.Xml.XmlTextReaderImpl.Throw(Exception e)
            at System.Xml.XmlTextReaderImpl.DtdParserProxy.System.Xml.IDtdParserAdapter.Throw(Exception e)
            at System.Xml.DtdParser.Throw(Int32 curPos, String res, String[] args)
            at System.Xml.DtdParser.ThrowUnexpectedToken(Int32 pos, String expectedToken1, String expectedToken2)
            at System.Xml.DtdParser.ParseExternalId(Token idTokenType, Token declType, String& publicId, String& systemId)
            at System.Xml.DtdParser.ParseInDocumentDtd(Boolean saveInternalSubset)
            at System.Xml.DtdParser.Parse(Boolean saveInternalSubset)
            at System.Xml.DtdParser.System.Xml.IDtdParser.ParseInternalDtd(IDtdParserAdapter adapter, Boolean saveInternalSubset)
            at System.Xml.XmlTextReaderImpl.ParseDtd()
            at System.Xml.XmlTextReaderImpl.ParseDoctypeDecl()
            at System.Xml.XmlTextReaderImpl.ParseDocumentContent()
            at System.Xml.XmlTextReaderImpl.Read()
            at System.Xml.Linq.XDocument.Load(XmlReader reader, LoadOptions options)
            at System.Xml.Linq.XDocument.Parse(String text, LoadOptions options)
            at System.Xml.Linq.XDocument.Parse(String text)
            at RestSharp.Deserializers.XmlDeserializer.Deserialize[T](IRestResponse response)
            at RestSharp.RestClient.Deserialize[T](IRestRequest request, IRestResponse raw)
       InnerException: 

```

**Comment by citrusO2** [comment]
looks like you have permission to retrieve the product but not to update it, there are several boxes of different permissions you can assign to an API-Key

(further, the error in the stacktrace seems to show you that the error message cannot be parsed correctly as well)

**Comment by MidnightBolt** [comment]
True that is why I can't get my head round it. You see, all boxes are checked. Look at the screenshot. Also, why is the additional 404. It does not make sense...
![webservice](https://cloud.githubusercontent.com/assets/6368662/6249240/292e5afc-b786-11e4-9733-ae07aa9f2222.png)

**Comment by citrusO2** [comment]
looks like the message comes from the apache2 server instead of prestashop itself. Maybe it's a problem with some .htaccess-file on the server
this could explain why prestasharp is crashing while trying to parse the response.

**Comment by MidnightBolt** [comment]
Thank you for your reply. It does indeed seem to be an error on my server. I made a blank .htaccess and regenerated it from the admin. It still does not work but I see a lot more directives in my .htaccess. So there was definitely something wrong there too. Now I filed a support request with my hosting partner. I will update this thread for future users to reference if they would have the same problem.
